common/src/main/java/io/druid/math/expr/Expr.java

```diff
@@ -258,320 +284,348 @@ public String toString()
   }
 }
 
-class BinMinusExpr extends BinaryOpExprBase
+abstract class BinaryNumericOpExprBase extends BinaryOpExprBase
 {
-
-  BinMinusExpr(String op, Expr left, Expr right)
+  public BinaryNumericOpExprBase(String op, Expr left, Expr right)
   {
     super(op, left, right);
   }
 
   @Override
-  public Number eval(ObjectBinding bindings)
+  public ExprEval eval(ObjectBinding bindings)
   {
-    Number leftVal = left.eval(bindings);
-    Number rightVal = right.eval(bindings);
-    if (isLong(leftVal, rightVal)) {
-      return leftVal.longValue() - rightVal.longValue();
-    } else {
-      return leftVal.doubleValue() - rightVal.doubleValue();
+    ExprEval leftVal = left.eval(bindings);
+    ExprEval rightVal = right.eval(bindings);
```

Comment:
I think with how this all works, it makes sense to have a `type()` method rather than polymorphism, since with polymorphism the value types would need to know how to operate on each other and that could get messy.